src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs

```diff
@@ -87,16 +144,154 @@ internal HttpMessageHandler CreateHandler(string name)
 
             configure(builder);
 
-            return builder.Build();
+            // Wrap the handler so we can ensure the inner handler outlives the outer handler.
+            var handler = new LifetimeTrackingHttpMessageHandler(builder.Build());
+
+            // Note that we can't start the timer here. That would introduce a very very subtle race condition
+            // with very short expiry times. We need to wait until we've actually handed out the handler once
+            // to start the timer.
+            // 
+            // Otherwise it would be possible that we start the timer here, immediately expire it (very short
+            // timer) and then dispose it without ever creating a client. That would be bad. It's unlikely
+            // this would happen, just we want to be sure.
+            return new ActiveHandlerTrackingEntry(name, handler, options.HandlerLifetime);
 
             void Configure(HttpMessageHandlerBuilder b)
             {
-                var options = _optionsMonitor.Get(name);
                 for (var i = 0; i < options.HttpMessageHandlerBuilderActions.Count; i++)
                 {
                     options.HttpMessageHandlerBuilderActions[i](b);
                 }
             }
         }
+
+        // Internal for tests
+        internal void ExpiryTimer_Tick(object state)
+        {
+            var active = (ActiveHandlerTrackingEntry)state;
+            active.DisposeTimer();
+
+            // The timer callback should be the only one removing from the active collection. If we can't find
+            // our entry in the collection, then this is a bug.
+            var removed = _activeHandlers.TryRemove(active.Name, out var found);
+            Debug.Assert(!removed);
+            Debug.Assert(object.ReferenceEquals(active, found));
+
+            // At this point the handler is no longer 'active' and will not be handed out to any new clients.
+            // However we haven't dropped our strong reference to the handler, so we can't yet determine if
+            // there are still any other outstanding references (we know there is at least one).
+            //
+            // We use a different state object to track expired handlers. This allows any other thread that acquired
+            // the 'active' entry to use it without safety problems.
+            var expired = new ExpiredHandlerTrackingEntry(active);
+            _expiredHandlers.Enqueue(expired);
+
+            StartCleanupTimer();
+        }
+
+        // Internal so it can be overridden in tests
+        internal virtual void StartCleanupTimer()
+        {
+            lock (_cleanupTimerLock)
+            {
+                if (_cleanupTimer == null)
+                {
+                    _cleanupTimer = new Timer(_cleanupCallback, null, DefaultCleanupInterval, Timeout.InfiniteTimeSpan);
```

Comment:
note: I prefer this pattern with timers - basically setting an alarm that will go off once at some point in the future. Since there's nothing recurring, you don't have any state to dispose. You just clean up each timer as it goes off.

The BCL already optimizes for using lots of short-lived timers.

Reply:
Good to know. Might want to do something similar with timeouts in SignalR :)